src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs

```diff
@@ -3652,26 +3652,33 @@ private void EmitMatchCharacterClass(string charClass, bool caseInsensitive)
                     return;
                 }
                 else if (numChars == 4 &&
-                         (setChars[0] | 0x20) == setChars[1] &&
-                         (setChars[2] | 0x20) == setChars[3])
+                         DifferByOneBit(setChars[0], setChars[1], out int mask) &&
+                         DifferByOneBit(setChars[2], setChars[3], out int mask2))
                 {
-                    // ((ch | 0x20) == setChars[1])
+                    // ((ch | mask) == setChars[1] | mask2)
                     Ldloc(tempLocal);
-                    Ldc(0x20);
+                    Ldc(mask);
                     Or();
-                    Ldc(setChars[1]);
+                    Ldc(setChars[1] | mask);
                     Ceq();
 
-                    // ((ch | 0x20) == setChars[3])
+                    // ((ch | mask2) == setChars[3] | mask2)
                     Ldloc(tempLocal);
-                    Ldc(0x20);
+                    Ldc(mask2);
                     Or();
-                    Ldc(setChars[3]);
+                    Ldc(setChars[3] | mask2);
                     Ceq();
 
                     Or();
                     return;
                 }
+
+                // Determines whether the 'a' and 'b' values differ by only a single bit, setting that bit in 'mask'.
+                static bool DifferByOneBit(char a, char b, out int mask)
```

Comment:
NIT: is there any value of having these common local functions between compiled and src gen refactored so that there is only one source of truth for them in case they need fixes?

Reply:
Sure